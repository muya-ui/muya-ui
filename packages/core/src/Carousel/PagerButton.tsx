import React from 'react';
import styled, { css } from 'styled-components';

import { useEventCallback } from '@muya-ui/utils';
import {
  colorUtils,
  GuildUpIcon,
  GuildDownIcon,
  GuildLeftIcon,
  GuildRightIcon,
} from '@muya-ui/theme-light';

import memoForwardRef from '../utils/memoForwardRef';
import useTheme from '../utils/useTheme';
import { withThemeForStyled } from '../utils/withTheme';
import { ICarouselPagerButtonProps } from './types';

const BaseNode = styled.button``;

const PagerButtonPure = memoForwardRef<HTMLButtonElement, ICarouselPagerButtonProps>(
  (props, ref) => {
    const {
      size,
      disabled,
      shape,
      transparent,
      arrow = 'left',
      side,
      icon,
      onClick,
      ...otherProps
    } = props;
    let iconNode = icon;
    const theme = useTheme();
    if (!iconNode) {
      const arrowIcon = {
        top: GuildUpIcon,
        bottom: GuildDownIcon,
        left: GuildLeftIcon,
        right: GuildRightIcon,
        ...theme.components.Carousel.PagerButton.arrowIcon,
      };
      const Icon = arrowIcon[arrow];
      iconNode = <Icon />;
    }
    const handleClick = useEventCallback(
      (e: React.MouseEvent<HTMLButtonElement, MouseEvent>) => {
        if (!disabled) {
          onClick && onClick(e);
        } else {
          e.preventDefault();
        }
      },
      [disabled],
    );

    return (
      <BaseNode {...otherProps} disabled={disabled} onClick={handleClick} ref={ref}>
        {iconNode}
      </BaseNode>
    );
  },
);

const PagerButton = styled(PagerButtonPure)<ICarouselPagerButtonProps>`
  /* 相对定位 */
  position: relative;

  /* flex + inline */
  display: inline-flex;

  /* 内部元素水平&垂直居中 */
  align-items: center;
  justify-content: center;
  cursor: pointer;
  user-select: none;
  outline: 0;
  margin: 0;
  border: none;
  font-family: inherit;
  text-transform: none;
  overflow: visible;
  box-sizing: border-box;
  appearance: button;
  ${props => {
    const {
      size = 'm',
      shape = 'rect',
      transparent,
      disabled,
      arrow = 'left',
      side,
      theme,
    } = props;
    const innerSide = side || arrow;
    const { opacity } = theme;
    const token = theme.components.Carousel.PagerButton;
    const width = token.width[size];
    const height = token.height[size];
    const iconSize = token.iconSize[size];
    let shapeCss;
    if (shape === 'rect') {
      const isH = arrow === 'left' || arrow === 'right';
      const innerWidth = isH ? width : height;
      const innerHeight = isH ? height : width;

      let sideCss;
      if (innerSide === 'left') {
        sideCss = css`
          border-top-left-radius: 0;
          border-bottom-left-radius: 0;
        `;
      } else if (innerSide === 'right') {
        sideCss = css`
          border-top-right-radius: 0;
          border-bottom-right-radius: 0;
        `;
      } else if (innerSide === 'top') {
        sideCss = css`
          border-top-right-radius: 0;
          border-top-left-radius: 0;
        `;
      } else {
        sideCss = css`
          border-bottom-right-radius: 0;
          border-bottom-left-radius: 0;
        `;
      }
      shapeCss = css`
        width: ${innerWidth}px;
        height: ${innerHeight}px;
        border-radius: ${token.borderRadius};
        ${sideCss}
      `;
    } else {
      shapeCss = css`
        width: ${height}px;
        height: ${height}px;
        border-radius: 50%;
      `;
    }
    let statusCss;
    if (disabled) {
      const color = colorUtils.transparentize(opacity.pattern.disabled, token.iconColor);
      statusCss = css`
        color: ${color};
        cursor: not-allowed;
      `;
    } else if (transparent) {
      statusCss = css`
        color: ${token.iconColor};
      `;
    } else {
      const bgAHColor = colorUtils.transparentize(token.opacity, token.bgColor);
      statusCss = css`
        color: ${token.iconColor};
        &:hover,
        &:active {
          background-color: ${bgAHColor};
        }
      `;
    }
    let bgCss;
    if (!transparent) {
      const bgColor = colorUtils.transparentize(token.hoverOpacity, token.bgColor);
      bgCss = css`
        background-color: ${bgColor};
      `;
    } else {
      bgCss = css`
        background-color: transparent;
      `;
    }

    return css`
      font-size: ${iconSize}px;
      ${shapeCss}
      ${bgCss}
      ${statusCss}
    `;
  }}
`;

export default withThemeForStyled(PagerButton);
