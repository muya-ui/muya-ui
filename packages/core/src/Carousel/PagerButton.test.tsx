import { mount } from 'enzyme';
import React from 'react';
import renderer from 'react-test-renderer';
import * as sinon from 'sinon';

import PagerButton from './PagerButton';

test('测试 PagerButton 正常情况', () => {
  const tree = renderer
    .create(
      <>
        <PagerButton arrow="left" />
        <PagerButton arrow="right" />
        <PagerButton arrow="top" />
        <PagerButton arrow="bottom" />
      </>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test('测试 PagerButton disabled', () => {
  const tree = renderer.create(<PagerButton disabled />).toJSON();
  expect(tree).toMatchSnapshot();
  expect(tree).toHaveStyleRule('cursor', 'not-allowed');
});
test('测试 PagerButton shape circle', () => {
  const tree = renderer.create(<PagerButton shape="circle" />).toJSON();
  expect(tree).toMatchSnapshot();
  expect(tree).toHaveStyleRule('border-radius', '50%');
});

test('测试 PagerButton transparent', () => {
  const tree = renderer.create(<PagerButton transparent />).toJSON();
  expect(tree).toMatchSnapshot();
});

test('测试 PagerButton onClick disabled', () => {
  const onClick = sinon.spy();
  const wrapper = mount(<PagerButton disabled onClick={onClick} />);
  wrapper.simulate('click');
  expect(() => {
    sinon.assert.notCalled(onClick);
  }).not.toThrow();
});

test('测试 PagerButton onClick', () => {
  const onClick = sinon.spy();
  const wrapper = mount(<PagerButton onClick={onClick} />);
  wrapper.simulate('click');
  expect(() => {
    sinon.assert.calledOnce(onClick);
  }).not.toThrow();
});
