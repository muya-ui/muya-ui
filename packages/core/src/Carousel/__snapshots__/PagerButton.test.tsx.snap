// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`测试 PagerButton disabled 1`] = `
.c1 {
  display: inline-block;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  font-style: normal;
  line-height: 0;
  text-align: center;
  text-transform: none;
  vertical-align: -0.125em;
  text-rendering: optimizeLegibility;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  width: 1em;
  height: 1em;
  overflow: hidden;
  fill: currentColor;
}

.c0 {
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  outline: 0;
  margin: 0;
  border: none;
  font-family: inherit;
  text-transform: none;
  overflow: visible;
  box-sizing: border-box;
  -webkit-appearance: button;
  -moz-appearance: button;
  appearance: button;
  font-size: 14px;
  width: 20px;
  height: 32px;
  border-radius: 2px;
  border-top-left-radius: 0;
  border-bottom-left-radius: 0;
  background-color: rgba(0,0,0,0.5);
  color: rgba(255,255,255,0.3);
  cursor: not-allowed;
}

<button
  className="c0"
  disabled={true}
  onClick={[Function]}
>
  <svg
    aria-hidden="false"
    className="c1"
    focusable="false"
    viewBox="0 0 1024 1024"
  >
    <path
      d="M501.725 499.81l260.036-381.385a21.67 21.67 0 00-5.742-30.121L626 3.575a21.67 21.67 0 00-29.796 5.851l-333.93 490.385a21.67 21.67 0 000 24.378l333.93 490.276a21.67 21.67 0 0029.796 5.96l129.476-84.512a21.67 21.67 0 006.284-30.013L501.725 524.514a21.67 21.67 0 010-24.703z"
    />
  </svg>
</button>
`;

exports[`测试 PagerButton shape circle 1`] = `
.c1 {
  display: inline-block;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  font-style: normal;
  line-height: 0;
  text-align: center;
  text-transform: none;
  vertical-align: -0.125em;
  text-rendering: optimizeLegibility;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  width: 1em;
  height: 1em;
  overflow: hidden;
  fill: currentColor;
}

.c0 {
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  outline: 0;
  margin: 0;
  border: none;
  font-family: inherit;
  text-transform: none;
  overflow: visible;
  box-sizing: border-box;
  -webkit-appearance: button;
  -moz-appearance: button;
  appearance: button;
  font-size: 14px;
  width: 32px;
  height: 32px;
  border-radius: 50%;
  background-color: rgba(0,0,0,0.5);
  color: rgba(255,255,255,1);
}

.c0:hover,
.c0:active {
  background-color: rgba(0,0,0,0.3);
}

<button
  className="c0"
  onClick={[Function]}
>
  <svg
    aria-hidden="false"
    className="c1"
    focusable="false"
    viewBox="0 0 1024 1024"
  >
    <path
      d="M501.725 499.81l260.036-381.385a21.67 21.67 0 00-5.742-30.121L626 3.575a21.67 21.67 0 00-29.796 5.851l-333.93 490.385a21.67 21.67 0 000 24.378l333.93 490.276a21.67 21.67 0 0029.796 5.96l129.476-84.512a21.67 21.67 0 006.284-30.013L501.725 524.514a21.67 21.67 0 010-24.703z"
    />
  </svg>
</button>
`;

exports[`测试 PagerButton transparent 1`] = `
.c1 {
  display: inline-block;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  font-style: normal;
  line-height: 0;
  text-align: center;
  text-transform: none;
  vertical-align: -0.125em;
  text-rendering: optimizeLegibility;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  width: 1em;
  height: 1em;
  overflow: hidden;
  fill: currentColor;
}

.c0 {
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  outline: 0;
  margin: 0;
  border: none;
  font-family: inherit;
  text-transform: none;
  overflow: visible;
  box-sizing: border-box;
  -webkit-appearance: button;
  -moz-appearance: button;
  appearance: button;
  font-size: 14px;
  width: 20px;
  height: 32px;
  border-radius: 2px;
  border-top-left-radius: 0;
  border-bottom-left-radius: 0;
  background-color: transparent;
  color: rgba(255,255,255,1);
}

<button
  className="c0"
  onClick={[Function]}
>
  <svg
    aria-hidden="false"
    className="c1"
    focusable="false"
    viewBox="0 0 1024 1024"
  >
    <path
      d="M501.725 499.81l260.036-381.385a21.67 21.67 0 00-5.742-30.121L626 3.575a21.67 21.67 0 00-29.796 5.851l-333.93 490.385a21.67 21.67 0 000 24.378l333.93 490.276a21.67 21.67 0 0029.796 5.96l129.476-84.512a21.67 21.67 0 006.284-30.013L501.725 524.514a21.67 21.67 0 010-24.703z"
    />
  </svg>
</button>
`;

exports[`测试 PagerButton 正常情况 1`] = `
Array [
  .c1 {
  display: inline-block;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  font-style: normal;
  line-height: 0;
  text-align: center;
  text-transform: none;
  vertical-align: -0.125em;
  text-rendering: optimizeLegibility;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  width: 1em;
  height: 1em;
  overflow: hidden;
  fill: currentColor;
}

.c0 {
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  outline: 0;
  margin: 0;
  border: none;
  font-family: inherit;
  text-transform: none;
  overflow: visible;
  box-sizing: border-box;
  -webkit-appearance: button;
  -moz-appearance: button;
  appearance: button;
  font-size: 14px;
  width: 20px;
  height: 32px;
  border-radius: 2px;
  border-top-left-radius: 0;
  border-bottom-left-radius: 0;
  background-color: rgba(0,0,0,0.5);
  color: rgba(255,255,255,1);
}

.c0:hover,
.c0:active {
  background-color: rgba(0,0,0,0.3);
}

<button
    className="c0"
    onClick={[Function]}
  >
    <svg
      aria-hidden="false"
      className="c1"
      focusable="false"
      viewBox="0 0 1024 1024"
    >
      <path
        d="M501.725 499.81l260.036-381.385a21.67 21.67 0 00-5.742-30.121L626 3.575a21.67 21.67 0 00-29.796 5.851l-333.93 490.385a21.67 21.67 0 000 24.378l333.93 490.276a21.67 21.67 0 0029.796 5.96l129.476-84.512a21.67 21.67 0 006.284-30.013L501.725 524.514a21.67 21.67 0 010-24.703z"
      />
    </svg>
  </button>,
  .c1 {
  display: inline-block;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  font-style: normal;
  line-height: 0;
  text-align: center;
  text-transform: none;
  vertical-align: -0.125em;
  text-rendering: optimizeLegibility;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  width: 1em;
  height: 1em;
  overflow: hidden;
  fill: currentColor;
}

.c0 {
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  outline: 0;
  margin: 0;
  border: none;
  font-family: inherit;
  text-transform: none;
  overflow: visible;
  box-sizing: border-box;
  -webkit-appearance: button;
  -moz-appearance: button;
  appearance: button;
  font-size: 14px;
  width: 20px;
  height: 32px;
  border-radius: 2px;
  border-top-right-radius: 0;
  border-bottom-right-radius: 0;
  background-color: rgba(0,0,0,0.5);
  color: rgba(255,255,255,1);
}

.c0:hover,
.c0:active {
  background-color: rgba(0,0,0,0.3);
}

<button
    className="c0"
    onClick={[Function]}
  >
    <svg
      aria-hidden="false"
      className="c1"
      focusable="false"
      viewBox="0 0 1024 1024"
    >
      <path
        d="M522.187 499.81L262.15 118.426a21.67 21.67 0 015.959-30.121L398.128 3.792a21.67 21.67 0 0129.796 5.851l333.82 490.168a21.67 21.67 0 010 24.378l-333.929 490.276a21.67 21.67 0 01-29.795 5.96l-129.477-84.512a21.67 21.67 0 01-6.284-30.013l260.036-381.386a21.67 21.67 0 00-.108-24.703z"
      />
    </svg>
  </button>,
  .c1 {
  display: inline-block;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  font-style: normal;
  line-height: 0;
  text-align: center;
  text-transform: none;
  vertical-align: -0.125em;
  text-rendering: optimizeLegibility;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  width: 1em;
  height: 1em;
  overflow: hidden;
  fill: currentColor;
}

.c0 {
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  outline: 0;
  margin: 0;
  border: none;
  font-family: inherit;
  text-transform: none;
  overflow: visible;
  box-sizing: border-box;
  -webkit-appearance: button;
  -moz-appearance: button;
  appearance: button;
  font-size: 14px;
  width: 32px;
  height: 20px;
  border-radius: 2px;
  border-top-right-radius: 0;
  border-top-left-radius: 0;
  background-color: rgba(0,0,0,0.5);
  color: rgba(255,255,255,1);
}

.c0:hover,
.c0:active {
  background-color: rgba(0,0,0,0.3);
}

<button
    className="c0"
    onClick={[Function]}
  >
    <svg
      aria-hidden="false"
      className="c1"
      focusable="false"
      viewBox="0 0 1024 1024"
    >
      <path
        d="M524.01 457.323L899.5 713.088a21.333 21.333 0 0029.866-5.973l83.2-127.52a21.333 21.333 0 00-5.845-29.291l-482.71-328.79a21.333 21.333 0 00-24.02 0L17.28 550.295a21.333 21.333 0 00-5.856 29.29l83.2 127.51a21.333 21.333 0 0029.877 5.973L499.99 457.333a21.333 21.333 0 0124.022 0z"
      />
    </svg>
  </button>,
  .c1 {
  display: inline-block;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  font-style: normal;
  line-height: 0;
  text-align: center;
  text-transform: none;
  vertical-align: -0.125em;
  text-rendering: optimizeLegibility;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  width: 1em;
  height: 1em;
  overflow: hidden;
  fill: currentColor;
}

.c0 {
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  outline: 0;
  margin: 0;
  border: none;
  font-family: inherit;
  text-transform: none;
  overflow: visible;
  box-sizing: border-box;
  -webkit-appearance: button;
  -moz-appearance: button;
  appearance: button;
  font-size: 14px;
  width: 32px;
  height: 20px;
  border-radius: 2px;
  border-bottom-right-radius: 0;
  border-bottom-left-radius: 0;
  background-color: rgba(0,0,0,0.5);
  color: rgba(255,255,255,1);
}

.c0:hover,
.c0:active {
  background-color: rgba(0,0,0,0.3);
}

<button
    className="c0"
    onClick={[Function]}
  >
    <svg
      aria-hidden="false"
      className="c1"
      focusable="false"
      viewBox="0 0 1024 1024"
    >
      <path
        d="M499.99 566.677L124.5 310.912a21.333 21.333 0 00-29.866 5.973l-83.2 127.52a21.333 21.333 0 005.845 29.291l482.71 328.79a21.333 21.333 0 0024.02 0l482.71-328.78a21.333 21.333 0 005.856-29.29l-83.2-127.51a21.333 21.333 0 00-29.877-5.973L524.01 566.667a21.333 21.333 0 01-24.022 0z"
      />
    </svg>
  </button>,
]
`;
